### PR TITLE
Replace Shell::WaitForFirstFrame with an asynchronous API that matches the Shell threading model

### PR DESCRIPTION
**Pull request by jason-simmons**
The Shell::WaitForFirstFrame API would block until the engine renders the first frame.  This could potentially cause issues if the shell is destroyed while a thread is waiting within a call to WaitForFirstFrame.

This PR replaces WaitForFirstFrame with an AddFirstFrameCallback API that invokes a callback when the first frame has been rendered.  If the shell is destroyed before the first frame, then the callback will never be executed.

See https://github.com/flutter/flutter/pull/190132

### CODE REVIEW

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 1]
```diff
@@ -612,10 +612,6 @@ - (void)notifyViewControllerDeallocated {
 }
 
 - (void)destroyContext {
-  // Clear any tasks waiting on first frame prior to destroying _shell.
-  if (_shell) {
-    _shell->CancelWaitForFirstFrame();
-  }
```
Comment (cbracken) [inline engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:618]: Thinking back to the issue this was added for, what happens if we have someone waiting? It we destroy the engine while we're waiting, I'm assuming we'll block for up to 3 seconds on the main thread?

I _think_ we can probably delete `_firstFrameWaiters` and the wait below altogether, then the waiter would just stick around in the background until timeout.

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 2]
```diff
@@ -1571,6 +1567,14 @@ - (void)onStatusBarTap {
 
 - (void)waitForFirstFrame:(NSTimeInterval)timeout
                  callback:(void (^_Nonnull)(BOOL didTimeout))callback {
+  auto first_frame_event = std::make_shared<fml::AutoResetWaitableEvent>();
+
+  self.shell.AddFirstFrameCallback([weak_event = std::weak_ptr(first_frame_event)] {
```
Comment (cbracken) [inline engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: self.shell has an FML_DCHECK in it so I think if we call this before FlutterEngine run or during shutdown, the assertion would fail (or null dereference in non-unopt builds), so we should check shell.

https://github.com/flutter/flutter/blob/f5f6313e8cce24472ebe25db99ae7a57c19f0fe0/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L416-L419

Not sure what the right value to set `didTimeOut` is in that case... `YES` seems more like what I'd expect since `NO` suggests the first frame rendered and someone might act on that.

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [inline engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: nit: With the change in destroyContext, nothing is cancelling these guys anymore so we should update the comment.

File: engine/src/flutter/shell/common/shell.h
Comment (cbracken) [inline engine/src/flutter/shell/common/shell.h]: This can be deleted now.

File: engine/src/flutter/shell/common/shell.h [hunk 3]
```diff
@@ -529,14 +525,8 @@ class Shell final : public PlatformView::Delegate,
   // waiting_for_first_frame_mutex_ in WaitForFirstFrame.
   std::atomic<bool> waiting_for_first_frame_ = true;
```
Comment (cbracken) [inline engine/src/flutter/shell/common/shell.h:516]: s/WaitForFirstFrame/AddFirstFrameCallback/

File: engine/src/flutter/shell/common/shell.h [hunk 4]
```diff
@@ -529,14 +525,8 @@ class Shell final : public PlatformView::Delegate,
...
 
-  // True when WaitForFirstFrame has been cancelled because the shell is
-  // shutting down and waiting threads should be unblocked.
-  //
-  // Guarded by waiting_for_first_frame_mutex_.
-  bool wait_for_first_frame_cancelled_ = false;
-
   std::mutex waiting_for_first_frame_mutex_;
-  std::condition_variable waiting_for_first_frame_condition_;
+  std::vector<std::function<void()>> waiting_for_first_frame_callbacks_;
```
Comment (cbracken) [inline engine/src/flutter/shell/common/shell.h:520]: I'd be tempted to leave the "guarded by waiting_for_first_frame_mutex_" comment to make it easier for future us to put two and two together.

File: engine/src/flutter/shell/common/shell.cc [hunk 5]
```diff
@@ -1446,7 +1447,14 @@ void Shell::OnAnimatorDraw(std::shared_ptr<FramePipeline> pipeline) {
 
           if (waiting_for_first_frame.load()) {
             waiting_for_first_frame.store(false);
-            waiting_for_first_frame_condition.notify_all();
+            std::vector<std::function<void()>> callbacks;
+            {
+              std::scoped_lock lock(waiting_for_first_frame_mutex);
+              std::swap(waiting_for_first_frame_callbacks, callbacks);
```
Comment (cbracken) [inline engine/src/flutter/shell/common/shell.cc:1460]: Wonder if it would be useful to leave a comment here about the relationship between waiting_for_first_frame and waiting_for_first_frame_callbacks and specifically the ordering guarantees these need.

The fact that the atomic bool is outside the lock but the callbacks vector is inside it got me stopping and thinking for longer than I'd like to admit; specifically what could happen in between. I wonder if we should just move the bool inside the lock to avoid anyone having to think about it.

I'm pretty convinced it's fine as-is, if someone registers in the gap they'll see it's false and invoke immediately.

File: engine/src/flutter/shell/common/shell.h [hunk 6]
```diff
@@ -339,18 +339,14 @@ class Shell final : public PlatformView::Delegate,
...
+  ///                       thread when the first frame is presented.  If the
+  ///                       first frame has already been presented, then the
+  ///                       callback will be invoked immediately.
   ///
-  /// @return     'kOk' when the first frame has been presented before the
-  ///             timeout successfully, 'kFailedPrecondition' if called from the
-  ///             GPU or UI thread, 'kDeadlineExceeded' if there is a timeout.
-  ///
-  fml::Status WaitForFirstFrame(fml::TimeDelta timeout);
+  void AddFirstFrameCallback(std::function<void()> callback);
```
Comment (cbracken) [inline engine/src/flutter/shell/common/shell.h]: We should probably state that any callbacks pending on teardown will never be invoked. Maybe worth mentioning that the callback may run on the raster thread so needs to be cheap.

consistency nit: The next frame callback uses fml::closure, which IIRC is just std::function<void()>. Doesn't bother me one way or another but noticed, since in my hacked-up attempt at this I went the other way. Up to you.

File: engine/src/flutter/shell/common/shell_unittests.cc [hunk 7]
```diff
@@ -1713,75 +1717,15 @@ TEST_F(ShellTest, WaitForFirstFrameTimeout) {
...
+  bool timeout = first_frame_event.WaitWithTimeout(fml::TimeDelta::Zero());
+  EXPECT_TRUE(timeout);
 
   DestroyShell(std::move(shell));
 }
 
-// Ensure CancelWaitForFirstFrame() correctly causes all tasks blocked on
-// WaitForFirstFrame() to return kAborted.
-//
-// See: b/521830222
```
Comment (cbracken) [inline engine/src/flutter/shell/common/shell_unittests.cc:1726]: Since this is the only test coverage we had for this bug, we should probably replace it with something roughly equivalent that verifies safe shutdown. Something where we register a callback, destroy the shell, and asserting the closure never ran.

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 8]
```diff
@@ -1583,18 +1587,10 @@ - (void)waitForFirstFrame:(NSTimeInterval)timeout
...
-      dispatch_group_leave(firstFrameWaiters);
-      return;
-    }
-
     fml::TimeDelta waitTime = fml::TimeDelta::FromMilliseconds(timeout * 1000);
-    fml::Status status = strongSelf.shell.WaitForFirstFrame(waitTime);
-    didTimeout = status.code() == fml::StatusCode::kDeadlineExceeded;
+    didTimeout = first_frame_event->WaitWithTimeout(waitTime);
     dispatch_group_leave(firstFrameWaiters);
   });
```
Comment (cbracken) [inline engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: I wonder if we could push this even further into async territory. What if when we register, we immediately kick off a timeout handler to fire on the main thread _AND_ we register the completion handler (which will get called on the raster thread but post the callback back to the main thread) these guys then race each other and the winner grabs the completion handler to a local, nils it out to the other can't fire it, then invokes it. 

The nice thing with this is we avoid the whole `_firstFrameWaiters` dispatch group and therefore avoid blocking the main thread for the timeout in `destroyContext`.

```objc
- (void)waitForFirstFrame:(NSTimeInterval)timeout
                 callback:(void (^_Nonnull)(BOOL didTimeout))callback {
  // Set up a completion handler that will nil itself out when it fires.
  __block void (^completion)(BOOL) = [callback copy];
  void (^complete)(BOOL) = ^(BOOL didTimeout) {
    if (completion) {
      void (^cb)(BOOL) = completion;
      completion = nil;
      cb(didTimeout);
    }
  };

  if (!_shell) {
    // No shell. Bail out since there'll never be a first frame.
    dispatch_async(dispatch_get_main_queue(), ^{
      complete(YES);
    });
    return;
  }

  // Register the first-frame callback and fire the timeout completion handler.
  // Both are scheduled on the main thread to guarantee ordering.
  // The winner nils out the completion handler so the loser is a no-op.
  _shell->AddFirstFrameCallback([complete] {
    dispatch_async(dispatch_get_main_queue(), ^{
      complete(NO);
    });
  });

  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(timeout * NSEC_PER_SEC)),
                 dispatch_get_main_queue(), ^{
                   complete(YES);
                 });
}
```

**Review by cbracken — COMMENTED** [review]
Overall this looks really good. 

Part of this is really similar to the hacked up local version I had. I had added a `FlutterFirstFramewaiter` class that tried to handle the cancellation relatively similarly to how it happens now (and cycling through them to cancel them all in destroyContext), but poking at this a bit more I think there might be a nicer solution that removes the need for all that and gets rid of the blocking (see my comment at the bottom).

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 1 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:618]: Removed `_firstFrameWaiters` and replaced its usage with the implementation in https://github.com/flutter/flutter/pull/191841#discussion_r3892893310

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 2 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: Done (using https://github.com/flutter/flutter/pull/191841#discussion_r3892893310)

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (jason-simmons) [reply engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: Removed this

File: engine/src/flutter/shell/common/shell.h
Comment (jason-simmons) [reply engine/src/flutter/shell/common/shell.h]: Removed this

File: engine/src/flutter/shell/common/shell.h (hunk 3 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/common/shell.h:516]: Done

File: engine/src/flutter/shell/common/shell.h (hunk 4 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/common/shell.h:520]: Done

File: engine/src/flutter/shell/common/shell.cc (hunk 5 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/common/shell.cc:1460]: Yeah - I considered changing `waiting_for_first_frame_` to a simple bool that is guarded by `waiting_for_first_frame_mutex_`.  But given that
`Shell::OnAnimatorDraw` is called on every frame, I decided to keep the existing atomic bool `waiting_for_first_frame_` to minimize the per-frame cost.

Added a comment and moved the `waiting_for_first_frame.store(false)` write into the `OnAnimatorDraw` block that holds the `waiting_for_first_frame_mutex_`.  That should make this clearer.

This works because `waiting_for_first_frame_` is set to true when the shell launches, but after the first frame it is set to false and remains false until the shell is suspended and resumed.
Users of `waiting_for_first_frame_callbacks_` will also access
`waiting_for_first_frame_` while holding the `waiting_for_first_frame_mutex_`.  But `OnAnimatorDraw` can check whether `waiting_for_first_frame_` has become `false`
without acquiring the mutex.

File: engine/src/flutter/shell/common/shell.h (hunk 6 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/common/shell.h]: Added a comment.

I considered using `fml::closure`, but the recent trend in the engine seems to be towards replacing usage of FML APIs with standard library or Abseil APIs.

File: engine/src/flutter/shell/common/shell_unittests.cc (hunk 7 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/common/shell_unittests.cc:1726]: Added this test.

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 8 above)
Comment (jason-simmons) [reply engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: Yes - that looks much cleaner!  Replaced the timeout mechanism in `FlutterEngine.waitForFirstFrame` with this implementation.

File: engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 9]
```diff
@@ -1571,50 +1560,37 @@ - (void)onStatusBarTap {
...
 
-    fml::TimeDelta waitTime = fml::TimeDelta::FromMilliseconds(timeout * 1000);
-    fml::Status status = strongSelf.shell.WaitForFirstFrame(waitTime);
-    didTimeout = status.code() == fml::StatusCode::kDeadlineExceeded;
-    dispatch_group_leave(firstFrameWaiters);
-  });
+  if (!_shell) {
+    // No shell. Bail out since there'll never be a first frame.
+    dispatch_async(dispatch_get_main_queue(), ^{
+      complete(YES);
```
[the remark this answers is not in the document]
Comment (jason-simmons) [reply engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:1576]: This comment is referring to an obsolete version of the PR.  `_firstFrameWaiters` and `first_frame_event` have been removed.

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

I patched this in locally and created an app with a configurable `RenderBinding.deferFirstFrame` so I could test delivery before the timeout (~2900ms), after it (5000ms), and right at the boundary (2950 ~ 2990 ms). I stuck a `.well-known/apple-app-site-association` on a personal domain for testing this.

Everything looks good. I also tested the existing version and one thing I noticed is this performs a little better right at the boundary; I actually had to push the first frame delay by a few milliseconds to get it to timeout with the new code.

This is a really nice improvement. Thanks so much for landing this.
